x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/search/WeightedTokensQueryBuilder.java

```diff
@@ -27,43 +22,60 @@
 import org.elasticsearch.xcontent.ParseField;
 import org.elasticsearch.xcontent.XContentBuilder;
 import org.elasticsearch.xcontent.XContentParser;
-import org.elasticsearch.xpack.core.ml.inference.results.TextExpansionResults.WeightedToken;
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
-
-import static org.elasticsearch.xpack.ml.queries.TextExpansionQueryBuilder.AllowedFieldType;
-import static org.elasticsearch.xpack.ml.queries.TextExpansionQueryBuilder.PRUNING_CONFIG;
+import java.util.stream.Collectors;
 
 public class WeightedTokensQueryBuilder extends AbstractQueryBuilder<WeightedTokensQueryBuilder> {
     public static final String NAME = "weighted_tokens";
 
     public static final ParseField TOKENS_FIELD = new ParseField("tokens");
+    public static final ParseField PRUNING_CONFIG = new ParseField("pruning_config");
     private final String fieldName;
     private final List<WeightedToken> tokens;
     @Nullable
     private final TokenPruningConfig tokenPruningConfig;
 
+    private final List<String> DEFAULT_ALLOWED_FIELD_TYPES = List.of("sparse_vector", "rank_features");
+    private final List<String> allowedFieldTypes;
+
     public WeightedTokensQueryBuilder(String fieldName, List<WeightedToken> tokens) {
         this(fieldName, tokens, null);
     }
 
     public WeightedTokensQueryBuilder(String fieldName, List<WeightedToken> tokens, @Nullable TokenPruningConfig tokenPruningConfig) {
+        this(fieldName, tokens, tokenPruningConfig, null);
+    }
+
+    public WeightedTokensQueryBuilder(
+        String fieldName,
+        List<WeightedToken> tokens,
+        @Nullable TokenPruningConfig tokenPruningConfig,
+        List<String> allowedFieldTypes
+    ) {
         this.fieldName = Objects.requireNonNull(fieldName, "[" + NAME + "] requires a fieldName");
         this.tokens = Objects.requireNonNull(tokens, "[" + NAME + "] requires tokens");
         if (tokens.isEmpty()) {
             throw new IllegalArgumentException("[" + NAME + "] requires at least one token");
         }
         this.tokenPruningConfig = tokenPruningConfig;
+        this.allowedFieldTypes = allowedFieldTypes != null ? allowedFieldTypes : DEFAULT_ALLOWED_FIELD_TYPES;
     }
 
     public WeightedTokensQueryBuilder(StreamInput in) throws IOException {
         super(in);
         this.fieldName = in.readString();
         this.tokens = in.readCollectionAsList(WeightedToken::new);
         this.tokenPruningConfig = in.readOptionalWriteable(TokenPruningConfig::new);
+        if (in.getTransportVersion().onOrAfter(TransportVersions.SPARSE_VECTOR_QUERY_ADDED)) {
+            this.allowedFieldTypes = in.readOptionalStringCollectionAsList();
+        } else {
+            this.allowedFieldTypes = DEFAULT_ALLOWED_FIELD_TYPES;
+        }
```

Comment:
I would prefer us not have this. Instead we should just have SparseVectorQueryBuilder be its own thing from the get go. Its ok to keep this around for internal clients. But since we don't want SparseVectorQueryBuilder to use it at all, I don't see the reason for updating it so that SparseVectorQueryBuilder can use it.